libarchive/archive_read_support_format_xar.c

```diff
@@ -3316,8 +3319,10 @@ expat_read_toc(struct archive_read *a)
 
 		d = NULL;
 		r = rd_contents(a, &d, &outbytes, &used, xar->toc_remaining);
-		if (r != ARCHIVE_OK)
+		if (r != ARCHIVE_OK) {
```

Comment:
I'm not sure about this.  This is certainly right if rd_contents returns ARCHIVE_FATAL, but if it returns something less serious, should we really release the parser here?

Reply:
In line 3340 it's released in every case as well. The variable `parser` is local and allocated in line 3302. I'd assume that it cannot be used outside and should be released.

Reply:
At least that's how I understand the code. I don't mind being corrected. In fact, it would be better to get this right at the first attempt.

Reply:
Yes, you're right.  Since `parser` is local, we need to be certain to clean it up here.  Good catch!